File: ios/calorietracker/Models/AddMenuConfig.swift
import Foundation

struct AddMenuGroupConfig: Codable, Equatable, Identifiable, Sendable {
    var id: UUID
    var name: String
    var methods: [FoodLogMethod]

    init(id: UUID = UUID(), name: String, methods: [FoodLogMethod]) {
        self.id = id
        self.name = name
        self.methods = methods
    }

    enum CodingKeys: String, CodingKey {
        case id, name, methods
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decodeIfPresent(UUID.self, forKey: .id) ?? UUID()
        name = try container.decode(String.self, forKey: .name)
        let rawMethods = try container.decode([String].self, forKey: .methods)
        methods = rawMethods.compactMap(FoodLogMethod.init(rawValue:))
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(name, forKey: .name)
        try container.encode(methods.map(\.rawValue), forKey: .methods)
    }
}

struct AddMenuConfig: Codable, Equatable, Sendable {
    static let currentVersion = 1
    static let storageKey = "addMenu.config"

    var version: Int
    var groups: [AddMenuGroupConfig]
    /// Used when `groups` is empty — flat list under + with no nesting.
    var flatMethods: [FoodLogMethod]

    init(version: Int = currentVersion, groups: [AddMenuGroupConfig], flatMethods: [FoodLogMethod] = []) {
        self.version = version
        self.groups = groups
        self.flatMethods = flatMethods
    }

    enum CodingKeys: String, CodingKey {
        case version, groups, flatMethods
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        version = try container.decodeIfPresent(Int.self, forKey: .version) ?? Self.currentVersion
        groups = try container.decodeIfPresent([AddMenuGroupConfig].self, forKey: .groups) ?? []
        let rawFlat = try container.decodeIfPresent([String].self, forKey: .flatMethods) ?? []
        flatMethods = rawFlat.compactMap(FoodLogMethod.init(rawValue:))
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(version, forKey: .version)
        try container.encode(groups, forKey: .groups)
        try container.encode(flatMethods.map(\.rawValue), forKey: .flatMethods)
    }

    /// Matches the pre-customization iOS Home + food menu.
    static let iOSDefault = AddMenuConfig(
        groups: [
            AddMenuGroupConfig(
                name: String(localized: "Reuse Meal", comment: "Default add menu group"),
                methods: [.copyFromDay, .favorites, .frequent, .recent]
            ),
            AddMenuGroupConfig(
                name: String(localized: "Describe Meal", comment: "Default add menu group"),
                methods: [.manual, .siriPhrases, .voice, .text]
            ),
            AddMenuGroupConfig(
                name: String(localized: "Photo & Scan", comment: "Default add menu group"),
                methods: [.barcode, .photos, .camera]
            ),
        ]
    )

    var usesFlatLayout: Bool { groups.isEmpty }

    var visibleMethods: [FoodLogMethod] {
        if usesFlatLayout {
            return flatMethods
        }
        return groups.flatMap(\.methods)
    }

    func sanitized() -> AddMenuConfig {
        let allowed = Set(FoodLogMethod.addMenuCases)
        var seen = Set<FoodLogMethod>()

        func filterMethods(_ methods: [FoodLogMethod]) -> [FoodLogMethod] {
            methods.compactMap { method in
                guard allowed.contains(method), !seen.contains(method) else { return nil }
                seen.insert(method)
                return method
            }
        }

        let trimmedGroups: [AddMenuGroupConfig] = groups.prefix(3).compactMap { group in
            let methods = filterMethods(group.methods)
            guard !methods.isEmpty else { return nil }
            return AddMenuGroupConfig(
                id: group.id,
                name: group.name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                    ? String(localized: "Group", comment: "Fallback add menu group name")
                    : group.name.trimmingCharacters(in: .whitespacesAndNewlines),
                methods: methods
            )
        }

        if trimmedGroups.isEmpty {
            return AddMenuConfig(
                version: Self.currentVersion,
                groups: [],
                flatMethods: filterMethods(flatMethods)
            )
        }

        return AddMenuConfig(
            version: Self.currentVersion,
            groups: trimmedGroups,
            flatMethods: []
        )
    }
}

enum AddMenuSettings {
    static func load(store: UserDefaults = .standard) -> AddMenuConfig {
        guard let data = store.data(forKey: AddMenuConfig.storageKey),
              let decoded = try? JSONDecoder().decode(AddMenuConfig.self, from: data)
        else {
            return .iOSDefault
        }
        let sanitized = decoded.sanitized()
        if sanitized.groups.isEmpty, sanitized.flatMethods.isEmpty {
            let hadConfiguredContent = decoded.groups.contains { !$0.methods.isEmpty }
                || !decoded.flatMethods.isEmpty
            if hadConfiguredContent {
                return .iOSDefault
            }
        }
        return sanitized
    }

    static func save(_ config: AddMenuConfig, store: UserDefaults = .standard) {
        let sanitized = config.sanitized()
        guard let data = try? JSONEncoder().encode(sanitized) else { return }
        store.set(data, forKey: AddMenuConfig.storageKey)
    }

    static func reset(store: UserDefaults = .standard) {
        store.removeObject(forKey: AddMenuConfig.storageKey)
    }
}

File: android/app/src/main/java/com/ayuvo/health/models/AddMenuConfig.kt
package com.ayuvo.health.models

import androidx.compose.runtime.Composable
import androidx.compose.ui.res.stringResource
import com.ayuvo.health.R
import kotlinx.serialization.Serializable
import java.util.UUID

@Serializable
data class AddMenuGroupConfig(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val methods: List<String>
)

@Serializable
data class AddMenuConfig(
    val version: Int = CURRENT_VERSION,
    val groups: List<AddMenuGroupConfig> = emptyList(),
    val flatMethods: List<String> = emptyList()
) {
    val usesFlatLayout: Boolean get() = groups.isEmpty()

    fun sanitized(): AddMenuConfig {
        val seen = mutableSetOf<FoodLogMethod>()

        fun filterMethods(raw: List<String>): List<String> {
            val result = mutableListOf<String>()
            for (key in raw) {
                val method = FoodLogMethod.fromStorage(key) ?: continue
                if (method !in seen) {
                    seen += method
                    result += method.storageKey
                }
            }
            return result
        }

        val trimmedGroups = groups.take(3).mapNotNull { group ->
            val methods = filterMethods(group.methods)
            if (methods.isEmpty()) return@mapNotNull null
            AddMenuGroupConfig(
                id = group.id.ifBlank { UUID.randomUUID().toString() },
                name = group.name.trim().ifBlank { DEFAULT_GROUP_FALLBACK_NAME },
                methods = methods
            )
        }

        return if (trimmedGroups.isEmpty()) {
            copy(version = CURRENT_VERSION, groups = emptyList(), flatMethods = filterMethods(flatMethods))
        } else {
            copy(version = CURRENT_VERSION, groups = trimmedGroups, flatMethods = emptyList())
        }
    }

    fun resolvedGroups(): List<ResolvedAddMenuGroup> =
        groups.mapNotNull { group ->
            val methods = group.methods.mapNotNull(FoodLogMethod::fromStorage)
            if (methods.isEmpty()) return@mapNotNull null
            ResolvedAddMenuGroup(
                id = group.id,
                name = group.name,
                methods = methods
            )
        }

    fun resolvedFlatMethods(): List<FoodLogMethod> =
        flatMethods.mapNotNull(FoodLogMethod::fromStorage)

    companion object {
        const val CURRENT_VERSION = 1
        const val STORAGE_KEY = "addMenu.config"

        /** Stable identifiers for default group names — resolved to localized strings at display time. */
        const val DEFAULT_GROUP_PHOTO_SCAN = "__default_photo_scan__"
        const val DEFAULT_GROUP_DESCRIBE_MEAL = "__default_describe_meal__"
        const val DEFAULT_GROUP_REUSE_MEAL = "__default_reuse_meal__"
        const val DEFAULT_GROUP_FALLBACK_NAME = "__default_group__"

        /** Matches the pre-customization Android Home + food menu. */
        val Default = AddMenuConfig(
            groups = listOf(
                AddMenuGroupConfig(
                    name = DEFAULT_GROUP_PHOTO_SCAN,
                    methods = listOf(
                        FoodLogMethod.CAMERA.storageKey,
                        FoodLogMethod.PHOTOS.storageKey,
                        FoodLogMethod.BARCODE.storageKey
                    )
                ),
                AddMenuGroupConfig(
                    name = DEFAULT_GROUP_DESCRIBE_MEAL,
                    methods = listOf(
                        FoodLogMethod.TEXT.storageKey,
                        FoodLogMethod.VOICE.storageKey,
                        FoodLogMethod.MANUAL.storageKey
                    )
                ),
                AddMenuGroupConfig(
                    name = DEFAULT_GROUP_REUSE_MEAL,
                    methods = listOf(
                        FoodLogMethod.RECENT.storageKey,
                        FoodLogMethod.FREQUENT.storageKey,
                        FoodLogMethod.FAVORITES.storageKey,
                        FoodLogMethod.COPY_FROM_DAY.storageKey
                    )
                )
            )
        )

        fun decode(raw: String?): AddMenuConfig {
            if (raw.isNullOrBlank()) return Default
            return runCatching {
                val decoded = json.decodeFromString<AddMenuConfig>(raw)
                val sanitized = decoded.sanitized()
                if (sanitized.groups.isEmpty() && sanitized.flatMethods.isEmpty()) {
                    val hadConfiguredContent = decoded.groups.any { it.methods.isNotEmpty() } ||
                        decoded.flatMethods.isNotEmpty()
                    if (hadConfiguredContent) Default else sanitized
                } else {
                    sanitized
                }
            }.getOrDefault(Default)
        }

        fun encode(config: AddMenuConfig): String =
            json.encodeToString(serializer(), config.sanitized())

        private val json = kotlinx.serialization.json.Json {
            ignoreUnknownKeys = true
            encodeDefaults = true
        }
    }
}

data class ResolvedAddMenuGroup(
    val id: String,
    val name: String,
    val methods: List<FoodLogMethod>
)

internal fun defaultGroupNameRes(storedName: String): Int? = when (storedName) {
    AddMenuConfig.DEFAULT_GROUP_PHOTO_SCAN -> R.string.home_menu_photo_scan
    AddMenuConfig.DEFAULT_GROUP_DESCRIBE_MEAL -> R.string.home_menu_describe_meal
    AddMenuConfig.DEFAULT_GROUP_REUSE_MEAL -> R.string.home_menu_reuse_meal
    AddMenuConfig.DEFAULT_GROUP_FALLBACK_NAME -> R.string.settings_add_menu_new_group
    else -> null
}

@Composable
fun AddMenuGroupConfig.displayName(): String {
    defaultGroupNameRes(name)?.let { return stringResource(it) }
    return name
}

@Composable
fun ResolvedAddMenuGroup.displayName(): String {
    defaultGroupNameRes(name)?.let { return stringResource(it) }
    return name
}
